### PR TITLE
[FLINK-2084] [core] Add an option to start Flink in streaming mode

### PR DESCRIPTION
**Pull request by StephanEwen**
The streaming mode subsumes the batch mode, as it allows to run both kinds of programs. Batch programs will most likely run slightly less efficient than in pure batch mode, due to lazy memory allocation (see below)

Adds new startup options to the JobManager and TaskManager.

Adds additional scripts `bin/start-cluster-streaming.sh` and `start-local-streaming.sh`.

Currently, streaming mode only sets the memory manager to lazy memory allocation to ensure heap is not blocked by batch memory manager

Includes [FLINK-2085] To support on-demand memory allocation in the memory manager.

### CODE REVIEW

File: flink-dist/src/main/flink-bin/bin/jobmanager.sh [hunk 1]
```diff
@@ -80,7 +81,7 @@ case $STARTSTOP in
         rotateLogFile $out
 
         echo "Starting Job Manager"
-        $JAVA_RUN $JVM_ARGS ${FLINK_ENV_JAVA_OPTS} "${log_setting[@]}" -classpath "`manglePathList "$FLINK_JM_CLASSPATH:$INTERNAL_HADOOP_CLASSPATHS"`" org.apache.flink.runtime.jobmanager.JobManager --executionMode $EXECUTIONMODE --configDir "$FLINK_CONF_DIR"  > "$out" 2>&1 < /dev/null &
+        $JAVA_RUN $JVM_ARGS ${FLINK_ENV_JAVA_OPTS} "${log_setting[@]}" -classpath "`manglePathList "$FLINK_JM_CLASSPATH:$INTERNAL_HADOOP_CLASSPATHS"`" org.apache.flink.runtime.jobmanager.JobManager --configDir "$FLINK_CONF_DIR" --executionMode $EXECUTIONMODE --streamingMode "$STREAMINGMODE" > "$out" 2>&1 < /dev/null &
```
Comment (rmetzger) [inline flink-dist/src/main/flink-bin/bin/jobmanager.sh]: With this change, the script now expects three arguments, right?
If you, you need to update this line as well
```
echo "Please specify 'start (cluster|local)' or stop"
```

**Comment by rmetzger** [comment]
+1 to merge.
The changes look good.

File: flink-runtime/src/test/java/org/apache/flink/runtime/memory/MemoryManagerLazyAllocationTest.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+			try {
+				this.memoryManager.allocatePages(mockInvoke, 1);
+				Assert.fail("Expected MemoryAllocationException.");
+			} catch (MemoryAllocationException maex) {
+				// expected
+			}
+			
+			Assert.assertTrue("The previously allocated segments were not valid any more.",
+																	allMemorySegmentsValid(segs));
+
```
Comment (mxm) [inline flink-runtime/src/test/java/org/apache/flink/runtime/memory/MemoryManagerLazyAllocationTest.java]: Formatting

**Comment by mxm** [comment]
+1 looks solid.

File: flink-dist/src/main/flink-bin/bin/jobmanager.sh (hunk 1 above)
Comment (StephanEwen) [inline flink-dist/src/main/flink-bin/bin/jobmanager.sh]: Actually, the streaming mode parameter is optional and defaults to "batch".

**Comment by StephanEwen** [comment]
Thank you for the comments.

I'll address them and merge this...

File: flink-dist/src/main/flink-bin/bin/jobmanager.sh (hunk 1 above)
Comment (rmetzger) [inline flink-dist/src/main/flink-bin/bin/jobmanager.sh]: Okay, sorry. I didn't see this.

Comment (StephanEwen) [inline flink-dist/src/main/flink-bin/bin/jobmanager.sh]: I'll add it anyways. It is a good hint.

**Pull request closed without merging**